### PR TITLE
Topology refactoring & fixes

### LINKED ISSUE #205: Topology optimization: move from prev scope into the fold scope outside the fold
This aqua
```
service OpH("op"):
  pr() -> []string
  pk(s: string)

func registerKeyPutValue(node_id: string) -> []string:
  nodes <- OpH.pr()
  for n <- nodes par:
    on n:
      OpH.pk(node_id)
  <- nodes
```
generates air code that will produce hops on every iteration
```
(fold nodes n
    (par
     (seq
      (call -relay- ("op" "noop") [])
      (xor
       (call n ("op" "puk") [node_id])
       (seq
        (call -relay- ("op" "noop") [])
        (call %init_peer_id% ("errorHandlingSrv" "error") [%last_error% 1])
       )
      )
     )
     (next n)
    )
   )
  )
```
But it is possible to optimize this and call `relay` before `fold`

### PR DESCRIPTION
**Pull request by alari**
Fixes #205

### CODE REVIEW

File: model/src/main/scala/aqua/model/func/raw/RawTag.scala [hunk 1]
```diff
@@ -67,64 +71,99 @@ case object ParTag extends ParGroupTag {
...
 }
 
-case object XorTag extends SeqGroupTag {
-  case object LeftBiased extends SeqGroupTag
+case object XorTag extends GroupTag {
+  case object LeftBiased extends GroupTag
+}
+
+case class XorParTag(xor: FuncOp, par: FuncOp) extends RawTag {
+  override def usesVarNames: Set[String] = xor.head.usesVarNames ++ par.head.usesVarNames
```
Comment (DieMyst) [inline model/src/main/scala/aqua/model/func/raw/RawTag.scala:79]: docs/why

File: model/src/main/scala/aqua/model/func/raw/RawTag.scala [hunk 2]
```diff
@@ -1,11 +1,13 @@
 package aqua.model.func.raw
 
 import aqua.model.ValueModel
+import aqua.model.ValueModel.varName
 import aqua.model.func.{Call, FuncModel}
 import cats.data.NonEmptyList
 import cats.data.Chain
 
 sealed trait RawTag {
+  def usesVarNames: Set[String] = Set.empty
```
Comment (DieMyst) [inline model/src/main/scala/aqua/model/func/raw/RawTag.scala:10]: docs

File: model/src/main/scala/aqua/model/func/raw/RawTag.scala [hunk 3]
```diff
@@ -67,64 +71,99 @@ case object ParTag extends ParGroupTag {
...
+                      values: NonEmptyList[ValueModel]
+                    ) extends NoExecTag
 
 object EmptyTag extends NoExecTag
 
 case class AbilityIdTag(
-  value: ValueModel,
-  service: String
-) extends NoExecTag
+                         value: ValueModel,
```
Comment (DieMyst) [inline model/src/main/scala/aqua/model/func/raw/RawTag.scala:144]: reformat

File: model/transform/src/main/scala/aqua/model/transform/cursor/ChainCursor.scala [hunk 4]
```diff
@@ -34,13 +34,14 @@ abstract class ChainCursor[C <: ChainCursor[C, T], T](make: NonEmptyList[ChainZi
...
-  def path: NonEmptyList[T] = tree.map(_.current)
+  lazy val path: NonEmptyList[T] = tree.map(_.current)
 
   // Move cursor up
+  // TODO: ensure this cursor's data is cached properly
   def moveUp: Option[C] = NonEmptyList.fromList(tree.tail).map(make)
 
   // Path to root, in form of Cursors; this is skipped
-  def pathToRoot: LazyList[C] = LazyList.unfold(this)(_.moveUp.map(c => c -> c))
+  val pathToRoot: LazyList[C] = LazyList.unfold(this)(_.moveUp.map(c => c -> c))
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/cursor/ChainCursor.scala:44]: maybe lazy too?

File: model/transform/src/main/scala/aqua/model/transform/topology/PathFinder.scala [hunk 5]
```diff
@@ -89,6 +49,25 @@ object PathFinder extends Logging {
     optimized
   }
 
+  def optimizePath(
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/PathFinder.scala:52]: docs

File: model/transform/src/main/scala/aqua/model/transform/topology/PathFinder.scala [hunk 6]
```diff
@@ -10,53 +10,13 @@ import scala.annotation.tailrec
...
-        case (acc, p) => acc :+ p
-      }
-    logger.trace(s"PEER IDS: $optimized")
-    logger.trace(s"PREFIX: $prefix")
-    logger.trace(s"SUFFIX: $suffix")
-    logger.trace(s"OPTIMIZED WITH PREFIX AND SUFFIX: $optimized")
-    val noPrefix = skipPrefix(optimized, prefix, optimized)
-    skipSuffix(noPrefix, suffix, noPrefix)
-  }
+  def findPath(fromOn: List[OnTag], toOn: List[OnTag]): Chain[ValueModel] =
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/PathFinder.scala:13]: docs. It is unclear why `path` is `Chain[ValueModel]`

File: model/transform/src/main/scala/aqua/model/transform/topology/RawCursor.scala [hunk 7]
```diff
@@ -2,21 +2,36 @@ package aqua.model.transform.topology
...
+import aqua.model.transform.cursor.*
+import cats.syntax.traverse.*
 import cats.free.Cofree
 import scribe.Logging
 
 // Can be heavily optimized by caching parent cursors, not just list of zippers
-case class RawCursor(tree: NonEmptyList[ChainZipper[FuncOp.Tree]])
-    extends ChainCursor[RawCursor, FuncOp.Tree](RawCursor.apply) with Logging {
+case class RawCursor(
+                      tree: NonEmptyList[ChainZipper[FuncOp.Tree]],
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/RawCursor.scala:15]: reformat

File: model/transform/src/main/scala/aqua/model/transform/topology/RawCursor.scala [hunk 8]
```diff
@@ -25,61 +40,28 @@ case class RawCursor(tree: NonEmptyList[ChainZipper[FuncOp.Tree]])
...
-   * Sequentially previous cursor
-   * @return
-   */
-  lazy val seqPrev: Option[RawCursor] =
-    parentTag.flatMap {
-      case p: SeqGroupTag if leftSiblings.nonEmpty =>
-        toPrevSibling.flatMap(c => c.lastExecuted orElse c.seqPrev)
-      case p =>
-        moveUp.flatMap(_.seqPrev)
+  lazy val isNoExec: Boolean =
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/RawCursor.scala:53]: docs

File: model/transform/src/main/scala/aqua/model/transform/topology/RawCursor.scala [hunk 9]
```diff
@@ -25,61 +40,28 @@ case class RawCursor(tree: NonEmptyList[ChainZipper[FuncOp.Tree]])
...
-  lazy val seqNext: Option[RawCursor] =
-    parentTag.flatMap {
-      case _: SeqGroupTag if rightSiblings.nonEmpty =>
-        toNextSibling.flatMap(c => c.firstExecuted orElse c.seqNext)
-      case _ => moveUp.flatMap(_.seqNext)
-    }
+  def hasExecLater: Boolean =
+    !allToRight.forall(_.isNoExec)
+
+  def exportsUsedLater: Boolean =
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/RawCursor.scala:63]: docs

File: model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala [hunk 10]
```diff
@@ -1,23 +1,388 @@
...
 import cats.Eval
-import cats.data.Chain.nil
+import cats.data.Chain.{==:, nil}
 import cats.data.{Chain, NonEmptyChain, NonEmptyList, OptionT}
 import cats.free.Cofree
 import cats.syntax.traverse.*
+import cats.syntax.apply.*
 import scribe.Logging
 
+case class Topology(
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:17]: docs

Comment (DieMyst) [reply model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:17]: with arguments explanation

File: model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala [hunk 11]
```diff
@@ -1,23 +1,388 @@
...
+  // Inside the left boundary of this element, what should be the scope
+  lazy val beginsOn: Eval[List[OnTag]] = begins.beginsOn(this).memoize
+
+  // After this element is done, what is the scope
+  lazy val endsOn: Eval[List[OnTag]] = ends.endsOn(this).memoize
+
+  // After this element is done, where should it move to prepare for the next one
+  lazy val afterOn: Eval[List[OnTag]] = after.afterOn(this).memoize
+
+  lazy val forceExit: Eval[Boolean] = after.forceExit(this).memoize
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:99]: docs

File: model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala [hunk 12]
```diff
@@ -1,23 +1,388 @@
...
+
+  // Where we finnaly are
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:101]: what does it mean? more clear explanation will be good

File: model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala [hunk 13]
```diff
@@ -1,23 +1,388 @@
...
+  lazy val pathAfter: Eval[Chain[ValueModel]] = after.pathAfter(this).memoize
+}
+
 object Topology extends Logging {
   type Tree = Cofree[Chain, RawTag]
   type Res = Cofree[Chain, ResolvedOp]
 
-  def resolve(op: Tree): Res = {
-    val resolved = resolveOnMoves(op).value
+  private def findAppendix(bef: List[OnTag], beg: List[OnTag]): Chain[ValueModel] =
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:113]: docs, what is appendix?

File: model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala [hunk 14]
```diff
@@ -1,23 +1,388 @@
...
+
+  object Default extends Before with Begins with Ends with After {
+    override def toString: String = "<default>"
+  }
+
+  // Parent == Seq, On
+  object SeqGroupBranch extends Before with After {
+    override def toString: String = "<seq>/*"
+
+    override def beforeOn(current: Topology): Eval[List[OnTag]] =
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:194]: some explanations in all branches will be great

Comment (DieMyst) [reply model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:194]: or maybe add a link to transform/index.md

File: model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala [hunk 15]
```diff
@@ -1,23 +1,388 @@
...
+
+    override def afterOn(current: Topology): Eval[List[OnTag]] = current.pathOn
+
+    override def forceExit(current: Topology): Eval[Boolean] = Eval.now(false)
+  }
+
+  object ParGroup extends Begins with Ends {
+    override def toString: String = "<par>"
+
+    override def beginsOn(current: Topology): Eval[List[OnTag]] =
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:282]: some explanation will be great

File: model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala [hunk 16]
```diff
@@ -1,23 +1,388 @@
...
+      current.children
+        .map(_.beginsOn.map(_.reverse))
+        .reduceLeftOption { case (b1e, b2e) =>
+          (b1e, b2e).mapN { case (b1, b2) =>
+            (b1 zip b2).takeWhile(_ == _).map(_._1)
+          }
+        }
+        .map(_.map(_.reverse)) getOrElse super.beginsOn(current)
+
+    override def endsOn(current: Topology): Eval[List[OnTag]] =
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:292]: some explanation will be great

File: model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala [hunk 17]
```diff
@@ -1,23 +1,388 @@
...
+        .map(_.flatMap {
+          case true => current.afterOn
+          case false => super.endsOn(current)
+        }) getOrElse super.endsOn(current)
+  }
+
+  object For extends Begins {
+    override def toString: String = "<for>"
+
+    override def beginsOn(current: Topology): Eval[List[OnTag]] =
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:307]: some explanation will be great

File: model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala [hunk 18]
```diff
@@ -1,23 +1,388 @@
...
+                // This is to take the outstanding relay and force moving there
+                (false, OnTag(r0, r) :: acc)
+              case ((true, acc), on) => (true, on :: acc)
+              case ((false, acc), _) => (false, acc)
+            }
+            ._2
+        )
+      } getOrElse super.beginsOn(current)
+
+    override def pathBefore(current: Topology): Eval[Chain[ValueModel]] =
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:326]: some explanation will be great

File: model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala [hunk 19]
```diff
@@ -1,23 +1,388 @@
...
+      ) getOrElse current.pathOn
+
+    final def finallyOn(current: Topology): Eval[List[OnTag]] =
+      current.forceExit.flatMap {
+        case true => current.afterOn
+        case false => current.endsOn
+      }
+
+    def pathAfter(current: Topology): Eval[Chain[ValueModel]] =
+      current.forceExit.flatMap {
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:178]: docs

File: model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala [hunk 20]
```diff
@@ -1,23 +1,388 @@
...
+
+    def afterOn(current: Topology): Eval[List[OnTag]] = current.pathOn
+
+    protected def afterParent(current: Topology): Eval[List[OnTag]] =
+      current.parent.map(
+        _.afterOn
+      ) getOrElse current.pathOn
+
+    final def finallyOn(current: Topology): Eval[List[OnTag]] =
+      current.forceExit.flatMap {
```
Comment (DieMyst) [inline model/transform/src/main/scala/aqua/model/transform/topology/Topology.scala:172]: docs

File: model/transform/src/main/scala/aqua/model/transform/cursor/ChainCursor.scala (hunk 4 above)
Comment (alari) [reply model/transform/src/main/scala/aqua/model/transform/cursor/ChainCursor.scala:44]: It's already lazy

**Review by DieMyst — APPROVED** [review]

**Pull request merged**